rust/src/ldap/ldap.rs

```diff
@@ -0,0 +1,417 @@
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&LdapTransaction> {
+        self.transactions.iter().find(|tx| tx.tx_id == tx_id + 1)
+    }
+
+    pub fn new_tx(&mut self) -> LdapTransaction {
+        let mut tx = LdapTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        if self.transactions.len() > unsafe { LDAP_MAX_TX } {
+            let mut index = self.tx_index_completed;
+            for tx_old in &mut self.transactions.range_mut(self.tx_index_completed..) {
+                index += 1;
+                if !tx_old.complete {
+                    tx_old.complete = true;
+                    tx.set_event(LdapEvent::TooManyTransactions);
+                    break;
+                }
+            }
+            self.tx_index_completed = index;
+        }
+        return tx;
+    }
+
+    fn find_request(&mut self, message_id: MessageID) -> Option<&mut LdapTransaction> {
+        self.transactions.iter_mut().find(|tx| {
+            tx.request
+                .as_ref()
+                .map_or(false, |req| req.message_id == message_id)
+        })
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> AppLayerResult {
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+
+        let mut start = input;
+        while !start.is_empty() {
+            match ldap_parse_msg(start) {
+                Ok((rem, msg)) => {
+                    start = rem;
+                    let mut tx = self.new_tx();
+                    let request = LdapMessage::from(msg);
+                    tx.complete = match request.protocol_op {
+                        ProtocolOp::UnbindRequest => true,
+                        _ => false,
+                    };
+                    tx.request = Some(request);
+                    self.transactions.push_back(tx);
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    let consumed = input.len() - start.len();
+                    let needed = start.len() + 1;
+                    return AppLayerResult::incomplete(consumed as u32, needed as u32);
+                }
+                Err(_) => {
+                    return AppLayerResult::err();
+                }
+            }
+        }
+
+        return AppLayerResult::ok();
+    }
+
+    fn parse_response(&mut self, input: &[u8]) -> AppLayerResult {
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+        let mut start = input;
+        while !start.is_empty() {
+            match ldap_parse_msg(start) {
+                Ok((rem, msg)) => {
+                    start = rem;
+
+                    let response = LdapMessage::from(msg);
+                    if let Some(tx) = self.find_request(response.message_id) {
```

Comment:
Id we do not find the associated request, we should still mark the tx complete, right ?

Reply:
Yes, and I think an event should be set.